source/cab/blocked-request-logger/blocked-request-logger.py

```diff
@@ -0,0 +1,413 @@
+			time.sleep(delay)
+		else:
+			break
+	else:
+		print('[get_web_acl_name] Failed ALL attempts to call API')
+
+	return web_acl_name
+
+#------------------------------------------------------------------------
+# Returns an array of all blocking rules associated with given web acl
+#------------------------------------------------------------------------
+def get_blocking_rule_ids(rules_info):
+	response = []
+	rule_ids = rules_info.keys()
+	for rule_id in rule_ids:
+		rule_name = rules_info[rule_id]['Name']
+		if rule_name.upper().find('WHITELIST') < 0:
+			response.append(rule_id)
+
+	# if need to add in Default_Action for testing purposes
+	# response.append('Default_Action')
+
+	return response
+
+#==========================
+# sampled-requests functions
+#==========================
+
+#------------------------------------------------------------------------
+# get-sampled-requests for given rule and request date/time -
+# need to think about how to use number of requests from which the
+# sample is drawn, since we can check if our sample is 'complete'!!
+#------------------------------------------------------------------------
+def get_sampled_requests(web_acl_id, rule_id, time_window, max_sampled_items):
+	response = None
+	print("getting sampled requests for rule_id = " + rule_id)
+	for attempt in range(API_CALL_NUM_RETRIES):
+		try:
+			response = waf.get_sampled_requests(WebAclId=web_acl_id, RuleId=rule_id, TimeWindow=time_window, MaxItems=max_sampled_items)
+		except Exception, e:
+			print(e)
+			delay = math.pow(2, attempt)
+			print('[waf_get_sampled_requests] Retrying in %d seconds...' % (delay))
+			time.sleep(delay)
+		else:
+			break
+	else:
+		print('[get_sampled_requests] Failed ALL attempts to call API')
+
+	return response
+
+#------------------------------------------------------------------------
+# check an individual sampled request against the blocked request
+#------------------------------------------------------------------------
+def sampled_request_matches_blocked_request(sampled_request, blocked_request):
+	sample_dt = sampled_request['Timestamp']
+	naive_trunc_sample_dt = aware_to_naive_datetime(sample_dt, truncated=True)
+	sample_ip = sampled_request['Request']['ClientIP']
+	sample_method = sampled_request['Request']['Method']
+	sample_uri = sampled_request['Request']['URI']
+
+	blocked_date = blocked_request['ReqDate']
+	blocked_time = blocked_request['ReqTime']
+	blocked_ip = blocked_request['ReqIP']
+	blocked_method = blocked_request['ReqMethod']
+	blocked_uri = blocked_request['ReqUri']
+
+	blocked_dt = get_datetime(blocked_date, blocked_time)
+	seconds_delta = datetime.timedelta(seconds=1)
+
+	return (blocked_dt == naive_trunc_sample_dt or blocked_dt + seconds_delta == naive_trunc_sample_dt or blocked_dt - seconds_delta == naive_trunc_sample_dt) and blocked_ip == sample_ip and blocked_method == sample_method # and sample_uri.startswith(blocked_uri)
```

Comment:
This line is too long, split the predicates in multiple lines.

Reply:
I tried to do that, but it wouldn't compile. Will take another look!